cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,293 @@
+        && (_CUDA_VMR::default_cuda_malloc_alignment % __alignment == 0);
+  }
+
+  //! @brief  Returns the default ``cudaMemPool_t`` from the current device
+  //! @throws cuda_error if retrieving the default ``cudaMemPool_t`` fails
+  //! @returns The default memory pool of the current device
+  _CCCL_NODISCARD static ::cudaMemPool_t __get_default_mem_pool(const int __device_id)
+  {
+    _LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
+
+    ::cudaMemPool_t __pool;
+    _CCCL_TRY_CUDA_API(
+      ::cudaDeviceGetDefaultMemPool, "Failed to call cudaDeviceGetDefaultMemPool", &__pool, __device_id);
+    return __pool;
+  }
+
+public:
+  //! @brief Default constructs the cuda_async_memory_resource using the default \c cudaMemPool_t of the current device
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource()
+      : __pool_(__get_default_mem_pool(_CUDA_VMR::__get_current_cuda_device()))
+  {}
+
+  //! @brief Constructs a cuda_async_memory_resource using the default \c cudaMemPool_t of a given deice
+  //! @throws cuda_error if retrieving the default \c cudaMemPool_t fails
+  cuda_async_memory_resource(::cuda::experimental::device_ref __device_ref)
+      : __pool_(__get_default_mem_pool(__device_ref.get()))
+  {}
+
+  cuda_async_memory_resource(int)                   = delete;
+  cuda_async_memory_resource(_CUDA_VSTD::nullptr_t) = delete;
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cudaMemPool_t.
+  //! @param __pool The \c cudaMemPool_t used to allocate memory.
+  explicit cuda_async_memory_resource(::cudaMemPool_t __pool) noexcept
+      : __pool_(__pool)
+  {}
+
+  //! @brief  Constructs the cuda_async_memory_resource from a \c cuda_memory_pool by calling native_handle()
+  //! @param __pool The \c cuda_memory_pool used to allocate memory.
+  explicit cuda_async_memory_resource(cuda_memory_pool& __pool) noexcept
+      : __pool_(__pool.native_handle())
+  {}
+
+  //! @brief Allocate device memory of size at least \p __bytes via cudaMallocFromPoolAsync.
+  //! @param __bytes The size in bytes of the allocation.
+  //! @param __alignment The requested alignment of the allocation.
+  //! @throws cuda::std::bad_alloc in case of invalid alignment, or
+  //! @throws cuda::cuda_error If an error code was return by the cuda api call.
+  //! @return Pointer to the newly allocated memory
+  _CCCL_NODISCARD void* allocate(const size_t __bytes,
+                                 const size_t __alignment = _CUDA_VMR::default_cuda_malloc_alignment)
+  {
+    if (!__is_valid_alignment(__alignment))
+    {
+      _CUDA_VSTD_NOVERSION::__throw_bad_alloc();
+    }
+
+    void* __ptr{nullptr};
+    _CCCL_TRY_CUDA_API(
+      ::cudaMallocFromPoolAsync,
+      "cuda_async_memory_resource::allocate failed to allocate with cudaMallocFromPoolAsync",
+      &__ptr,
+      __bytes,
+      __pool_,
+      ::cudaStream_t{0});
```

Comment:
Not sure about this thing, is a non-async allocate needed for the async resource?
Same in deallocate

Reply:
Yeah this is needed because an async resource must always be a resource too

Reply:
That said we should definitely use the current default stream

Reply:
I think this needs some different solution, right now this allocation will block waiting on all work on the current device (or just error out if current device is not set). Then as far as I understand there is no way to wait for this allocation, unless we synchronize here (which is even worse).
We probably need to use synchronous allocation not from pool or have some internal stream.

Reply:
We discussed this internally and the issue we face is that we do not know the device the pool has been created on, so there is no way of even creating a stream on the "right" device.

So for now we will keep this as is and will try to work out a way of extracting the device from the `cudaMemPool_t`